File: sdk/kotlin-enhanced/src/main/kotlin/com/flagent/enhanced/fetcher/SnapshotFetcher.kt
package com.flagent.enhanced.fetcher

import com.flagent.client.apis.ExportApi
import com.flagent.client.apis.FlagApi
import com.flagent.enhanced.model.*
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.add
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put

/**
 * SnapshotFetcher - fetches flag configurations from server.
 *
 * Provides bootstrap and refresh mechanisms for downloading flag snapshots
 * from the Flagent server.
 *
 * @param exportApi Export API client for bulk flag data
 * @param flagApi Flag API client for individual flag operations
 *
 * @example
 * ```
 * val fetcher = SnapshotFetcher(exportApi, flagApi)
 * val snapshot = fetcher.fetchSnapshot()
 * ```
 */
class SnapshotFetcher(
    private val exportApi: ExportApi,
    private val flagApi: FlagApi
) {
    
    /**
     * Fetch complete flag snapshot from server.
     *
     * Downloads all flags with their segments, variants, constraints, and distributions.
     * Uses the export API for efficient bulk data transfer.
     *
     * @param ttlMs TTL for the snapshot in milliseconds (default 5 minutes)
     * @return Flag snapshot ready for local evaluation
     * @throws Exception if fetch fails
     */
    suspend fun fetchSnapshot(ttlMs: Long = 300000): FlagSnapshot = withContext(Dispatchers.IO) {
        try {
            // Fetch snapshot from export API (more efficient)
            val response = exportApi.getExportEvalCacheJSON()
            val body = response.body()
            val json = mapToJsonObject(body)
            val flags = parseExportJson(json)
            
            FlagSnapshot(
                flags = flags,
                revision = generateRevision(),
                fetchedAt = System.currentTimeMillis(),
                ttlMs = ttlMs
            )
        } catch (e: Exception) {
            // Fallback to individual flag fetch if export fails
            fetchSnapshotFallback(ttlMs)
        }
    }

    /**
     * Fallback method to fetch flags individually.
     *
     * Used when export API is unavailable or fails.
     */
    private suspend fun fetchSnapshotFallback(ttlMs: Long): FlagSnapshot {
        val response = flagApi.findFlags(
            limit = 1000,
            offset = null,
            enabled = null,
            description = null,
            key = null,
            descriptionLike = null,
            preload = true,
            deleted = false,
            tags = null
        )
        val flagsList = response.body()
        
        val flags = flagsList.mapNotNull { flagDto ->
            val flagId = flagDto.id
            
            val segments = flagDto.segments?.map { segmentDto ->
                LocalSegment(
                    id = segmentDto.id,
                    rank = (segmentDto.rank).toInt(),
                    rolloutPercent = (segmentDto.rolloutPercent).toInt(),
                    constraints = segmentDto.constraints?.map { constraintDto ->
                        LocalConstraint(
                            id = constraintDto.id,
                            property = constraintDto.`property`,
                            operator = constraintDto.`operator`.value,
                            value = constraintDto.`value`
                        )
                    } ?: emptyList(),
                    distributions = segmentDto.distributions?.map { distDto ->
                        LocalDistribution(
                            id = distDto.id,
                            variantID = distDto.variantID,
                            variantKey = distDto.variantKey ?: "",
                            percent = distDto.percent.toInt()
                        )
                    } ?: emptyList(),
                    description = segmentDto.description
                )
            } ?: emptyList()
            
            val variants = flagDto.variants?.map { variantDto ->
                LocalVariant(
                    id = variantDto.id,
                    key = variantDto.key,
                    attachment = variantDto.attachment
                )
            } ?: emptyList()
            
            flagId to LocalFlag(
                id = flagId,
                key = flagDto.key,
                enabled = flagDto.enabled,
                segments = segments,
                variants = variants,
                description = flagDto.description,
                entityType = flagDto.entityType,
                updatedAt = null
            )
        }.toMap()
        
        return FlagSnapshot(
            flags = flags,
            revision = generateRevision(),
            fetchedAt = System.currentTimeMillis(),
            ttlMs = ttlMs
        )
    }

    /**
     * Parse export JSON format into flag map.
     *
     * The export format is optimized for evaluation cache.
     */
    private fun parseExportJson(json: JsonObject): Map<Long, LocalFlag> {
        // Export format structure varies, this is a basic parser
        // May need adjustment based on actual export format
        val flags = mutableMapOf<Long, LocalFlag>()
        
        try {
            json.forEach { (key, value) ->
                val flagId = key.toLongOrNull() ?: return@forEach
                val flagObj = value.jsonObject
                
                val segments = flagObj["segments"]?.jsonArray?.map { segmentEl ->
                    val segmentObj = segmentEl.jsonObject
                    LocalSegment(
                        id = segmentObj["id"]?.jsonPrimitive?.content?.toLongOrNull() ?: 0L,
                        rank = segmentObj["rank"]?.jsonPrimitive?.content?.toIntOrNull() ?: 0,
                        rolloutPercent = segmentObj["rolloutPercent"]?.jsonPrimitive?.content?.toIntOrNull() ?: 100,
                        constraints = segmentObj["constraints"]?.jsonArray?.map { constraintEl ->
                            val constraintObj = constraintEl.jsonObject
                            LocalConstraint(
                                id = constraintObj["id"]?.jsonPrimitive?.content?.toLongOrNull() ?: 0L,
                                property = constraintObj["property"]?.jsonPrimitive?.content ?: "",
                                operator = constraintObj["operator"]?.jsonPrimitive?.content ?: "EQ",
                                value = constraintObj["value"]?.jsonPrimitive?.content
                            )
                        } ?: emptyList(),
                        distributions = segmentObj["distributions"]?.jsonArray?.map { distEl ->
                            val distObj = distEl.jsonObject
                            LocalDistribution(
                                id = distObj["id"]?.jsonPrimitive?.content?.toLongOrNull() ?: 0L,
                                variantID = distObj["variantID"]?.jsonPrimitive?.content?.toLongOrNull() ?: 0L,
                                variantKey = distObj["variantKey"]?.jsonPrimitive?.content ?: "",
                                percent = distObj["percent"]?.jsonPrimitive?.content?.toIntOrNull() ?: 0
                            )
                        } ?: emptyList()
                    )
                } ?: emptyList()
                
                val variants = flagObj["variants"]?.jsonArray?.map { variantEl ->
                    val variantObj = variantEl.jsonObject
                    LocalVariant(
                        id = variantObj["id"]?.jsonPrimitive?.content?.toLongOrNull() ?: 0L,
                        key = variantObj["key"]?.jsonPrimitive?.content ?: "",
                        attachment = variantObj["attachment"]?.jsonObject
                    )
                } ?: emptyList()
                
                flags[flagId] = LocalFlag(
                    id = flagId,
                    key = flagObj["key"]?.jsonPrimitive?.content ?: "",
                    enabled = flagObj["enabled"]?.jsonPrimitive?.content?.toBooleanStrictOrNull() ?: false,
                    segments = segments,
                    variants = variants,
                    description = flagObj["description"]?.jsonPrimitive?.content,
                    entityType = flagObj["entityType"]?.jsonPrimitive?.content
                )
            }
        } catch (e: Exception) {
            // If parsing fails, return empty map (will trigger fallback)
            throw e
        }
        
        return flags
    }

    /**
     * Convert Map (from Export API response) to JsonObject for parsing.
     */
    private fun mapToJsonObject(map: Map<String, Any?>): JsonObject {
        return buildJsonObject {
            map.forEach { (k, v) ->
                put(k, anyToJsonElement(v))
            }
        }
    }

    private fun anyToJsonElement(v: Any?): JsonElement {
        if (v == null) return JsonPrimitive("")
        return when (v) {
            is Map<*, *> -> mapToJsonObject(v as Map<String, Any?>)
            is List<*> -> JsonArray(v.map { anyToJsonElement(it) })
            is Number -> JsonPrimitive(v)
            is Boolean -> JsonPrimitive(v)
            is String -> JsonPrimitive(v)
            else -> JsonPrimitive(v.toString())
        }
    }

    /**
     * Generate revision ID for snapshot.
     */
    private fun generateRevision(): String {
        return "snapshot-${System.currentTimeMillis()}"
    }

    /**
     * Fetch only flags that changed since last revision.
     *
     * @param lastRevision Last known revision
     * @param ttlMs TTL for the snapshot
     * @return Updated snapshot or null if no changes
     */
    suspend fun fetchDelta(lastRevision: String?, ttlMs: Long = 300000): FlagSnapshot? {
        // For now, always fetch full snapshot
        // TODO: Implement delta fetch when backend supports it
        return fetchSnapshot(ttlMs)
    }
}

File: sdk/swift-enhanced/Sources/FlagentEnhanced/Offline/SnapshotFetcher.swift
import Foundation
import FlagentClient
import AnyCodable
import Combine

/// Fetches snapshot from Export API (eval cache JSON) or falls back to FlagAPI.findFlags.
@available(macOS 10.15, iOS 13.0, *)
struct SnapshotFetcher {
    private let ttlMs: Int64

    init(ttlMs: Int64) {
        self.ttlMs = ttlMs
    }

    func fetchSnapshot() async throws -> OfflineFlagSnapshot {
        do {
            let body = try await ExportAPI.getExportEvalCacheJSON().async()
            let json = body
            let flags = parseExportJSON(json)
            return OfflineFlagSnapshot(
                flags: flags,
                revision: "snapshot-\(currentTimeMs())",
                fetchedAtMs: currentTimeMs(),
                ttlMs: ttlMs
            )
        } catch {
            return try await fetchSnapshotFallback()
        }
    }

    private func fetchSnapshotFallback() async throws -> OfflineFlagSnapshot {
        let flagsList = try await FlagAPI.findFlags(limit: 1000, offset: nil, enabled: nil, description: nil, key: nil, descriptionLike: nil, preload: true, deleted: false, tags: nil).async()
        var flags: [Int64: OfflineLocalFlag] = [:]
        for flagDto in flagsList {
            let segments = (flagDto.segments ?? []).map { seg -> OfflineLocalSegment in
                let constraints = (seg.constraints ?? []).map { c in
                    OfflineLocalConstraint(
                        id: c.id,
                        property: c.property,
                        operator_: c._operator.rawValue,
                        value: c.value
                    )
                }
                let distributions = (seg.distributions ?? []).map { d in
                    OfflineLocalDistribution(
                        id: d.id,
                        variantID: d.variantID,
                        percent: Int(d.percent)
                    )
                }
                return OfflineLocalSegment(
                    id: seg.id,
                    rank: Int(seg.rank),
                    rolloutPercent: Int(seg.rolloutPercent),
                    constraints: constraints,
                    distributions: distributions
                )
            }
            let variants = (flagDto.variants ?? []).map { v in
                OfflineLocalVariant(id: v.id, key: v.key, attachment: v.attachment)
            }
            let localFlag = OfflineLocalFlag(
                id: flagDto.id,
                key: flagDto.key,
                enabled: flagDto.enabled,
                segments: segments,
                variants: variants
            )
            flags[flagDto.id] = localFlag
        }
        return OfflineFlagSnapshot(
            flags: flags,
            revision: "snapshot-\(currentTimeMs())",
            fetchedAtMs: currentTimeMs(),
            ttlMs: ttlMs
        )
    }

    private func parseExportJSON(_ json: [String: AnyCodable]) -> [Int64: OfflineLocalFlag] {
        var flags: [Int64: OfflineLocalFlag] = [:]
        if let flagsArray = json["flags"]?.value as? [[String: AnyCodable]] {
            for flagEl in flagsArray {
                let flagObj = flagEl
                guard let flagId = int64(from: flagObj["id"]) else { continue }
                if let f = parseFlagObject(flagObj, flagId: flagId) {
                    flags[flagId] = f
                }
            }
        } else {
            for (key, value) in json {
                guard let flagId = Int64(key) else { continue }
                guard let flagObj = (value.value as? [String: AnyCodable]) else { continue }
                if let f = parseFlagObject(flagObj, flagId: flagId) {
                    flags[flagId] = f
                }
            }
        }
        return flags
    }

    private func parseFlagObject(_ flagObj: [String: AnyCodable], flagId: Int64) -> OfflineLocalFlag? {
        let segments: [OfflineLocalSegment] = {
            guard let arr = flagObj["segments"]?.value as? [[String: AnyCodable]] else { return [] }
            return arr.compactMap { segEl in
                let id = int64(from: segEl["id"]) ?? 0
                let rank = int(from: segEl["rank"]) ?? 0
                let rolloutPercent = int(from: segEl["rolloutPercent"]) ?? 100
                let constraints: [OfflineLocalConstraint] = {
                    guard let cArr = segEl["constraints"]?.value as? [[String: AnyCodable]] else { return [] }
                    return cArr.compactMap { c in
                        guard let cid = int64(from: c["id"]),
                              let prop = c["property"]?.value as? String,
                              let val = c["value"]?.value as? String else { return nil }
                        let op = c["operator"]?.value as? String ?? "EQ"
                        return OfflineLocalConstraint(id: cid, property: prop, operator_: op, value: val)
                    }
                }()
                let distributions: [OfflineLocalDistribution] = {
                    guard let dArr = segEl["distributions"]?.value as? [[String: AnyCodable]] else { return [] }
                    return dArr.compactMap { d in
                        guard let did = int64(from: d["id"]),
                              let vid = int64(from: d["variantId"]) ?? int64(from: d["variantID"]),
                              let pct = int(from: d["percent"]) else { return nil }
                        return OfflineLocalDistribution(id: did, variantID: vid, percent: pct)
                    }
                }()
                return OfflineLocalSegment(id: id, rank: rank, rolloutPercent: rolloutPercent, constraints: constraints, distributions: distributions)
            }
        }()
        let variants: [OfflineLocalVariant] = {
            guard let vArr = flagObj["variants"]?.value as? [[String: AnyCodable]] else { return [] }
            return vArr.compactMap { v in
                guard let vid = int64(from: v["id"]),
                      let key = v["key"]?.value as? String else { return nil }
                let attachment = v["attachment"]?.value as? [String: AnyCodable]
                return OfflineLocalVariant(id: vid, key: key, attachment: attachment)
            }
        }()
        let key = flagObj["key"]?.value as? String ?? ""
        let enabled = (flagObj["enabled"]?.value as? Bool) ?? false
        return OfflineLocalFlag(id: flagId, key: key, enabled: enabled, segments: segments, variants: variants)
    }

    private func int64(from ac: AnyCodable?) -> Int64? {
        guard let ac = ac else { return nil }
        if let n = ac.value as? Int { return Int64(n) }
        if let n = ac.value as? Int64 { return n }
        if let s = ac.value as? String { return Int64(s) }
        return nil
    }

    private func int(from ac: AnyCodable?) -> Int? {
        guard let ac = ac else { return nil }
        if let n = ac.value as? Int { return n }
        if let n = ac.value as? Int64 { return Int(n) }
        if let s = ac.value as? String { return Int(s) }
        return nil
    }

    private func currentTimeMs() -> Int64 {
        Int64(Date().timeIntervalSince1970 * 1000)
    }
}
